File: python/oaaf/constraints.py
"""Argument constraints and the subsumption matrix (AAT -01 section 4.5).

Implemented from the normative rules, not translated from the reference. Two
rules govern: the permitted (parent, derived) type pairs are closed-world (any
pair not explicitly permitted is rejected), and an unrecognized constraint type
is rejected rather than ignored.
"""

from __future__ import annotations

import json
from typing import Any

KNOWN_TYPES = {
    "exact",
    "range",
    "one_of",
    "not_one_of",
    "contains",
    "subset",
    "wildcard",
    "all",
    "any",
}


def is_constraint(value: Any) -> bool:
    if not isinstance(value, dict):
        return False
    t = value.get("constraint_type")
    if not isinstance(t, str) or t not in KNOWN_TYPES:
        return False
    if t == "one_of":
        return isinstance(value.get("values"), list)
    if t == "not_one_of":
        return isinstance(value.get("excluded"), list)
    if t == "contains":
        return isinstance(value.get("required"), list)
    if t == "subset":
        return isinstance(value.get("allowed"), list)
    if t in ("all", "any"):
        cs = value.get("constraints")
        return isinstance(cs, list) and all(is_constraint(c) for c in cs)
    if t == "exact":
        return "value" in value
    if t == "range":
        return (value.get("min") is None or _is_num(value["min"])) and (
            value.get("max") is None or _is_num(value["max"])
        )
    return True


def _is_num(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _json_equal(a: Any, b: Any) -> bool:
    return _stable(a) == _stable(b)


def _stable(v: Any) -> str:
    return json.dumps(v, sort_keys=True, separators=(",", ":"))


def _includes(lst: list[Any], value: Any) -> bool:
    return any(_json_equal(x, value) for x in lst)


def _subset_of(inner: list[Any], outer: list[Any]) -> bool:
    return all(_includes(outer, x) for x in inner)


def satisfies(constraint: dict, value: Any) -> bool:
    """Does a supplied argument value satisfy a constraint?"""
    t = constraint["constraint_type"]
    if t == "wildcard":
        return True
    if t == "exact":
        return _json_equal(constraint["value"], value)
    if t == "one_of":
        return _includes(constraint["values"], value)
    if t == "not_one_of":
        return not _includes(constraint["excluded"], value)
    if t == "range":
        return _satisfies_range(constraint, value)
    if t == "contains":
        return isinstance(value, list) and _subset_of(constraint["required"], value)
    if t == "subset":
        return isinstance(value, list) and _subset_of(value, constraint["allowed"])
    if t == "all":
        return all(satisfies(c, value) for c in constraint["constraints"])
    if t == "any":
        return any(satisfies(c, value) for c in constraint["constraints"])
    return False


def _satisfies_range(constraint: dict, value: Any) -> bool:
    if not _is_num(value):
        return False
    min_inc = constraint.get("min_inclusive", True)
    max_inc = constraint.get("max_inclusive", True)
    lo, hi = constraint.get("min"), constraint.get("max")
    if lo is not None and (value < lo if min_inc else value <= lo):
        return False
    if hi is not None and (value > hi if max_inc else value >= hi):
        return False
    return True


def is_permitted_pair(parent: str, derived: str) -> bool:
    """Permitted (parent, derived) type pairs, per section 4.5 (closed-world)."""
    if parent == "wildcard":
        return True
    if derived == "wildcard":
        return False
    if derived == "exact":
        return parent in ("exact", "range", "one_of", "wildcard")
    return parent == derived


def subsumes(parent: dict, derived: dict) -> bool:
    """True when a derived constraint is no broader than its parent."""
    if not is_constraint(parent) or not is_constraint(derived):
        return False
    if not is_permitted_pair(parent["constraint_type"], derived["constraint_type"]):
        return False
    if parent["constraint_type"] == "wildcard":
        return True

    dt = derived["constraint_type"]
    if dt == "exact":
        return _exact_subsumed_by(parent, derived["value"])
    if dt == "range":
        return parent["constraint_type"] == "range" and _range_subsumes(parent, derived)
    if dt == "one_of":
        return parent["constraint_type"] == "one_of" and _subset_of(
            derived["values"], parent["values"]
        )
    if dt == "not_one_of":
        return parent["constraint_type"] == "not_one_of" and _subset_of(
            parent["excluded"], derived["excluded"]
        )
    if dt == "contains":
        return parent["constraint_type"] == "contains" and _subset_of(
            parent["required"], derived["required"]
        )
    if dt == "subset":
        return parent["constraint_type"] == "subset" and _subset_of(
            derived["allowed"], parent["allowed"]
        )
    if dt == "all":
        return parent["constraint_type"] == "all" and _subsumes_all(parent, derived)
    if dt == "any":
        return parent["constraint_type"] == "any" and _subsumes_any(parent, derived)
    return False


def _exact_subsumed_by(parent: dict, value: Any) -> bool:
    t = parent["constraint_type"]
    if t == "exact":
        return _json_equal(parent["value"], value)
    if t == "range":
        return _satisfies_range(parent, value)
    if t == "one_of":
        return _includes(parent["values"], value)
    return False


def _range_subsumes(parent: dict, derived: dict) -> bool:
    p_min_inc = parent.get("min_inclusive", True)
    d_min_inc = derived.get("min_inclusive", True)
    p_max_inc = parent.get("max_inclusive", True)
    d_max_inc = derived.get("max_inclusive", True)
    if parent.get("min") is not None:
        if derived.get("min") is None or derived["min"] < parent["min"]:
            return False
        if derived["min"] == parent["min"] and d_min_inc and not p_min_inc:
            return False
    if parent.get("max") is not None:
        if derived.get("max") is None or derived["max"] > parent["max"]:
            return False
        if derived["max"] == parent["max"] and d_max_inc and not p_max_inc:
            return False
    return True


def _subsumes_all(parent: dict, derived: dict) -> bool:
    used: set[int] = set()

    def match(idx: int) -> bool:
        if idx == len(parent["constraints"]):
            return True
        p_clause = parent["constraints"][idx]
        for i, d_clause in enumerate(derived["constraints"]):
            if i in used or not subsumes(p_clause, d_clause):
                continue
            used.add(i)
            if match(idx + 1):
                return True
            used.discard(i)
        return False

    return match(0)


def _subsumes_any(parent: dict, derived: dict) -> bool:
    if len(derived["constraints"]) == 0:
        return False
    return all(
        any(subsumes(p, d) for p in parent["constraints"]) for d in derived["constraints"]
    )

File: packages/typescript/src/aat/constraints.ts
/**
 * Argument constraints and the subsumption matrix from AAT -01 §4.5.
 *
 * Two rules govern everything here.
 *
 * The matrix is **closed-world**: the draft states that any (parent type,
 * derived type) pair not explicitly permitted MUST be rejected. So this module
 * enumerates the permitted pairs and rejects everything else, including pairs
 * that look harmless.
 *
 * Unrecognized constraint types are rejected rather than ignored. A constraint
 * that cannot be understood cannot be shown to narrow anything.
 *
 * Two ambiguities in -01 are handled here by failing closed; both are recorded
 * in RFC-0001 and should be raised with the draft author. See PERMITTED_PAIRS
 * and `subsumesAny` for the specifics.
 */

export type ConstraintType =
  'exact' | 'range' | 'one_of' | 'not_one_of' | 'contains' | 'subset' | 'wildcard' | 'all' | 'any';

export interface ExactConstraint {
  constraint_type: 'exact';
  value: unknown;
}

export interface RangeConstraint {
  constraint_type: 'range';
  min?: number;
  max?: number;
  min_inclusive?: boolean;
  max_inclusive?: boolean;
}

export interface OneOfConstraint {
  constraint_type: 'one_of';
  values: unknown[];
}

export interface NotOneOfConstraint {
  constraint_type: 'not_one_of';
  excluded: unknown[];
}

export interface ContainsConstraint {
  constraint_type: 'contains';
  required: unknown[];
}

export interface SubsetConstraint {
  constraint_type: 'subset';
  allowed: unknown[];
}

export interface WildcardConstraint {
  constraint_type: 'wildcard';
}

export interface AllConstraint {
  constraint_type: 'all';
  constraints: Constraint[];
}

export interface AnyConstraint {
  constraint_type: 'any';
  constraints: Constraint[];
}

export type Constraint =
  | ExactConstraint
  | RangeConstraint
  | OneOfConstraint
  | NotOneOfConstraint
  | ContainsConstraint
  | SubsetConstraint
  | WildcardConstraint
  | AllConstraint
  | AnyConstraint;

const KNOWN_TYPES: ReadonlySet<string> = new Set<ConstraintType>([
  'exact',
  'range',
  'one_of',
  'not_one_of',
  'contains',
  'subset',
  'wildcard',
  'all',
  'any',
]);

/** True when `value` is a structurally valid constraint of a type this build knows. */
export function isConstraint(value: unknown): value is Constraint {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const type = (value as { constraint_type?: unknown }).constraint_type;
  if (typeof type !== 'string' || !KNOWN_TYPES.has(type)) return false;

  const c = value as Record<string, unknown>;
  switch (type) {
    case 'one_of':
      return Array.isArray(c['values']);
    case 'not_one_of':
      return Array.isArray(c['excluded']);
    case 'contains':
      return Array.isArray(c['required']);
    case 'subset':
      return Array.isArray(c['allowed']);
    case 'all':
    case 'any':
      return Array.isArray(c['constraints']) && c['constraints'].every(isConstraint);
    case 'exact':
      return 'value' in c;
    case 'range':
      return (
        (c['min'] === undefined || typeof c['min'] === 'number') &&
        (c['max'] === undefined || typeof c['max'] === 'number')
      );
    default:
      return true;
  }
}

/**
 * Structural equality by JSON value.
 *
 * Constraint values are JSON, so key order is not significant; sorting keys
 * before comparison avoids treating `{a:1,b:2}` and `{b:2,a:1}` as different
 * permitted values.
 */
function jsonEqual(a: unknown, b: unknown): boolean {
  return stableStringify(a) === stableStringify(b);
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') return JSON.stringify(value) ?? 'null';
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`;
  const entries = Object.entries(value as Record<string, unknown>).sort(([a], [b]) =>
    a < b ? -1 : a > b ? 1 : 0,
  );
  return `{${entries.map(([k, v]) => `${JSON.stringify(k)}:${stableStringify(v)}`).join(',')}}`;
}

function includesValue(list: readonly unknown[], value: unknown): boolean {
  return list.some((entry) => jsonEqual(entry, value));
}

function isSubsetOf(inner: readonly unknown[], outer: readonly unknown[]): boolean {
  return inner.every((entry) => includesValue(outer, entry));
}

// ---------------------------------------------------------------------------
// Evaluation: does an argument value satisfy a constraint?
// ---------------------------------------------------------------------------

/** Evaluate a supplied argument value against a constraint. */
export function satisfies(constraint: Constraint, value: unknown): boolean {
  switch (constraint.constraint_type) {
    case 'wildcard':
      return true;
    case 'exact':
      return jsonEqual(constraint.value, value);
    case 'one_of':
      return includesValue(constraint.values, value);
    case 'not_one_of':
      return !includesValue(constraint.excluded, value);
    case 'range':
      return satisfiesRange(constraint, value);
    case 'contains':
      return Array.isArray(value) && isSubsetOf(constraint.required, value);
    case 'subset':
      return Array.isArray(value) && isSubsetOf(value, constraint.allowed);
    case 'all':
      return constraint.constraints.every((inner) => satisfies(inner, value));
    case 'any':
      return constraint.constraints.some((inner) => satisfies(inner, value));
  }
}

function satisfiesRange(constraint: RangeConstraint, value: unknown): boolean {
  if (typeof value !== 'number' || !Number.isFinite(value)) return false;
  const minInclusive = constraint.min_inclusive ?? true;
  const maxInclusive = constraint.max_inclusive ?? true;
  if (constraint.min !== undefined) {
    if (minInclusive ? value < constraint.min : value <= constraint.min) return false;
  }
  if (constraint.max !== undefined) {
    if (maxInclusive ? value > constraint.max : value >= constraint.max) return false;
  }
  return true;
}

// ---------------------------------------------------------------------------
// Subsumption: is a derived constraint no broader than its parent?
// ---------------------------------------------------------------------------

/**
 * Permitted (parent type, derived type) pairs, transcribed from AAT -01 §4.5.
 *
 * Deliberately no more permissive than the draft's table. Two consequences are
 * surprising and are believed to be under-specification rather than intent:
 *
 *   - No pair lists `contains` or `subset` as a *parent* type, so a parent
 *     constraint of either kind cannot be restated in a derived token at all.
 *   - `wildcard` may only be narrowed to `exact` or `wildcard` — not to
 *     `one_of` or `range`, which would be the natural narrowings.
 *
 * Both are filed in RFC-0001 for upstream clarification. Until the draft says
 * otherwise, the closed-world rule is explicit and failing closed is never the
 * unsafe direction: a child may always omit the argument or the tool instead.
 */
const PERMITTED_PAIRS: ReadonlySet<string> = new Set([
  'exact>exact',
  'range>exact',
  'one_of>exact',
  'wildcard>exact',
  'range>range',
  'one_of>one_of',
  'not_one_of>not_one_of',
  'wildcard>wildcard',
  'all>all',
  'any>any',
]);

/** True when the (parent, derived) type pair appears in the draft's table. */
export function isPermittedPair(parent: ConstraintType, derived: ConstraintType): boolean {
  return PERMITTED_PAIRS.has(`${parent}>${derived}`);
}

/**
 * True when `derived` is no broader than `parent`.
 *
 * Rejects unknown types, unpermitted type pairs, and any case whose narrowing
 * cannot be demonstrated.
 */
export function subsumes(parent: Constraint, derived: Constraint): boolean {
  if (!isConstraint(parent) || !isConstraint(derived)) return false;
  if (!isPermittedPair(parent.constraint_type, derived.constraint_type)) return false;

  switch (parent.constraint_type) {
    case 'exact':
      return derived.constraint_type === 'exact' && jsonEqual(parent.value, derived.value);

    case 'wildcard':
      // Only 'exact' and 'wildcard' reach here; both are narrowings of wildcard.
      return true;

    case 'one_of':
      if (derived.constraint_type === 'exact') return includesValue(parent.values, derived.value);
      if (derived.constraint_type === 'one_of') return isSubsetOf(derived.values, parent.values);
      return false;

    case 'not_one_of':
      // Narrowing means excluding at least as much.
      return (
        derived.constraint_type === 'not_one_of' && isSubsetOf(parent.excluded, derived.excluded)
      );

    case 'range':
      if (derived.constraint_type === 'exact') return satisfiesRange(parent, derived.value);
      if (derived.constraint_type === 'range') return rangeSubsumes(parent, derived);
      return false;

    case 'all':
      return derived.constraint_type === 'all' && subsumesAll(parent, derived);

    case 'any':
      return derived.constraint_type === 'any' && subsumesAny(parent, derived);

    case 'contains':
    case 'subset':
      // No permitted pair names these as a parent type; unreachable via the
      // pair check above, and rejected here too rather than falling through.
      return false;
  }
}

/** A derived range must sit inside the parent range, inclusivity included. */
function rangeSubsumes(parent: RangeConstraint, derived: RangeConstraint): boolean {
  const parentMinInc = parent.min_inclusive ?? true;
  const derivedMinInc = derived.min_inclusive ?? true;
  const parentMaxInc = parent.max_inclusive ?? true;
  const derivedMaxInc = derived.max_inclusive ?? true;

  if (parent.min !== undefined) {
    // An unbounded derived lower end cannot be inside a bounded parent one.
    if (derived.min === undefined) return false;
    if (derived.min < parent.min) return false;
    // Equal bounds: derived must not be more permissive at the boundary.
    if (derived.min === parent.min && derivedMinInc && !parentMinInc) return false;
  }

  if (parent.max !== undefined) {
    if (derived.max === undefined) return false;
    if (derived.max > parent.max) return false;
    if (derived.max === parent.max && derivedMaxInc && !parentMaxInc) return false;
  }

  return true;
}

/**
 * Conjunction. The parent permits the intersection of its clauses, so every
 * parent clause must still be enforced by some derived clause. Extra derived
 * clauses only narrow further and are allowed.
 */
function subsumesAll(parent: AllConstraint, derived: AllConstraint): boolean {
  return parent.constraints.every((parentClause) =>
    derived.constraints.some((derivedClause) => subsumes(parentClause, derivedClause)),
  );
}

/**
 * Disjunction. The parent permits the union of its clauses, so narrowing holds
 * when every *derived* clause falls inside some parent clause.
 *
 * Descriptions of -01 §4.5 are not consistent about this direction. The
 * opposite reading — every parent clause subsumed by some derived clause —
 * would permit a derived token to add alternatives the parent never granted,
 * which is widening. The narrowing-preserving direction is implemented here and
 * the discrepancy is recorded in RFC-0001.
 */
function subsumesAny(parent: AnyConstraint, derived: AnyConstraint): boolean {
  return derived.constraints.every((derivedClause) =>
    parent.constraints.some((parentClause) => subsumes(parentClause, derivedClause)),
  );
}
